storage/zynq7000-sdcard/sdcard.c

```diff
@@ -0,0 +1,1212 @@
+
+static inline sdcard_hostData_t *sdcard_getHostForSlot(unsigned int slot)
+{
+	return (slot >= PLATFORM_SDIO_N_HOSTS) ? NULL : &sdio_hosts[slot];
+}
+
+#define GET_HOST_OR_RETURN(...) \
+	sdcard_hostData_t *host = sdcard_getHostForSlot(slot); \
+	if (host == NULL) { \
+		return __VA_ARGS__; \
+	}
+
+static int sdcard_configClockAndPower(sdcard_hostData_t *host, uint32_t freq);
+static int sdcard_wideAndFast(sdcard_hostData_t *host);
+
+
+/* Resets parts or all of the host according to the given reset type.
+ * Argument must be one of CLOCK_CONTROL_RESET_*
+ */
+static int sdhost_reset(sdcard_hostData_t *host, uint32_t resetType)
+{
+	*(host->base + SDHOST_REG_CLOCK_CONTROL) |= resetType;
+	for (int i = 0; i < SDHOST_RETRIES; ++i) {
+		usleep(10);
+		uint32_t val = *(host->base + SDHOST_REG_CLOCK_CONTROL) & resetType;
+		if (val == 0) {
+			return 0;
+		}
+	}
+
+	return -1;
+}
+
+
+static inline bool sdcard_isWriteProtected(sdcard_hostData_t *host)
+{
+	if (host->isWPPinSupported) {
+		uint32_t val = *(host->base + SDHOST_REG_PRES_STATE) & PRES_STATE_WRITE_PROT_PIN;
+		if (val == 0) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+
+static void *sdhost_mmapDMA(size_t size)
+{
+	size_t sz = (size + _PAGE_SIZE - 1) & ~(_PAGE_SIZE - 1);
+	void *p = mmap(NULL, sz, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS | MAP_UNCACHED, NULL, 0);
+	if (p == MAP_FAILED) {
+		return NULL;
+	}
+
+	return p;
+}
+
+
+/* Checks physical address of a buffer in virtual memory and checks if the buffer is contiguous. */
+static addr_t sdhost_mphys(void *v, size_t *vSize)
+{
+	addr_t phys = va2pa(v);
+	size_t sz = _PAGE_SIZE - (phys & (_PAGE_SIZE - 1));
+
+	while (sz < *vSize) {
+		addr_t nextPhys = va2pa(v + sz);
+		if (nextPhys != (phys + sz)) {
+			break;
+		}
+
+		sz += _PAGE_SIZE;
+	}
+
+	if (sz < *vSize) {
+		*vSize = sz;
+	}
+
+	return phys;
+}
+
+
+static int sdhost_allocDMA(sdcard_hostData_t *host)
+{
+	size_t psize = SDCARD_MAX_TRANSFER;
+
+	host->dmaBuffer = sdhost_mmapDMA(psize);
+	if (host->dmaBuffer == NULL) {
+		return -ENOMEM;
+	}
+
+	host->dmaBufferPhys = sdhost_mphys(host->dmaBuffer, &psize);
+
+	if (SDCARD_MAX_TRANSFER != psize) {
+		munmap(host->dmaBuffer, SDCARD_MAX_TRANSFER);
+		host->dmaBuffer = NULL;
+		return -1;
+	}
+
+	if ((host->dmaBufferPhys & 3) != 0) {
+		munmap(host->dmaBuffer, SDCARD_MAX_TRANSFER);
+		host->dmaBuffer = NULL;
+		return -1;
+	}
+
+	return 0;
+}
+
+
+static int sdhost_isr(unsigned int n, void *arg)
+{
+	/* SD Host controller interrupts are triggered by level, not by edge.
+	 * This means that ISR has to remove the reason for interrupt (zero out bit in STATUS register)
+	 * or disable the interrupt (zero out bit in SIGNAL_ENABLE register) - otherwise we get stuck
+	 * in infinite interrupt loop.
+	 */
+	sdcard_hostData_t *host = (sdcard_hostData_t *)arg;
+	uint32_t activeInterrupts = *(host->base + SDHOST_REG_INTR_STATUS);
+	*(host->base + SDHOST_REG_INTR_SIGNAL_ENABLE) &= ~activeInterrupts;
```

Comment:
Why go with the disabling the interrupt approach? This might cause some nasty race condition (e.g. when we need to modify `SDHOST_REG_INTR_SIGNAL_ENABLE` register not in the ISR - there's no way to synchronize access with the ISR)